3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,264 @@
+
+  /**
+   * Handle receipt of a message indicating that creative(s) have sent
+   * event(s) to this frame
+   * @param {!Array<!../src/iframe-transport-common.IframeTransportEvent>}
+   * events An array of events
+   * @private
+   */
+  processEventsMessage_(events) {
+    user().assert(events && events.length,
+        'Received empty events list in ' + this.win_.location.href);
+    this.win_.onNewAmpAnalyticsInstance =
+        this.win_.onNewAmpAnalyticsInstance || null;
+    user().assert(this.win_.onNewAmpAnalyticsInstance,
+        'Must implement onNewAmpAnalyticsInstance in ' +
+        this.win_.location.href);
+    events.forEach(event => {
+      const transportId = event['transportId'];
+      const message = event['message'];
+      try {
+        if (!this.creativeEventRouters_[transportId]) {
+          this.creativeEventRouters_[transportId] =
+              new CreativeEventRouter(
+                  this.win_, this.sentinel_, transportId);
+          try {
+            this.win_.onNewAmpAnalyticsInstance(
+                this.creativeEventRouters_[transportId]);
+          } catch (e) {
+            user().error(TAG_, 'Caught exception in' +
+              ' onNewAmpAnalyticsInstance: ' + e.message);
+            throw e;
+          }
+        }
+        this.creativeEventRouters_[transportId]
+            .sendMessageToListener(message);
+      } catch (e) {
+        user().error(TAG_, 'Failed to pass message to event listener: ' +
+          e.message);
+      }
+    });
+  }
+
+  /**
+   * Test method to ensure sentinel set correctly
+   * @returns {string}
+   * @VisibleForTesting
+   */
+  getSentinel() {
+    return this.sentinel_;
+  }
+
+  /**
+   * Gets the mapping of creative senderId to
+   * CreativeEventRouter
+   * @returns {!Object.<string, !CreativeEventRouter>}
+   * @VisibleForTesting
+   */
+  getCreativeEventRouters() {
+    return this.creativeEventRouters_;
+  }
+
+  /**
+   * Gets rid of the mapping to CreativeEventRouter
+   * @VisibleForTesting
+   */
+  reset() {
+    this.creativeEventRouters_ = {};
+  }
+
+  /**
+   * Takes the raw postMessage event, and extracts from it the actual data
+   * payload
+   * @param event
+   * @returns {JsonObject}
+   * @private
+   */
+  extractMessage_(event) {
+    user().assert(event, 'Received null event in ' + this.win_.name);
+    const data = String(getData(event));
+    user().assert(data, 'Received empty event in ' + this.win_.name);
+    let startIndex;
+    user().assert((startIndex = data.indexOf('-') + 1) > 0,
+        'Received truncated events message in ' + this.win_.name);
+    return tryParseJson(data.substr(startIndex)) || null;
+  }
+}
+
+if (!window.AMP_TEST) {
+  try {
+    new EventRouter(window);
+  } catch (e) {
+    user().error(TAG_, 'Failed to construct EventRouter: ' +
+      e.message);
+  }
+}
+
+/**
+ * Receives messages bound for this cross-domain iframe, from a particular
+ * creative.
+ */
+export class CreativeEventRouter {
```

Comment:
Currently CreativeEventRouter provides an association between a creative (transport) and the customer-provided listener function, and EventRouter handles communication with the parent frame (for any/all transports). Therefore the function passed to registerCreativeEventListener() is specific to one creative.